### ISSUE TITLE
Async code restarts interceptor chain, which runs code twice and causes errors in transactions

### ISSUE THREAD

**Issue by ballad89** [pedestal/pedestal#661]
# Description

## Expected Behavior
After an async operation, I expect the interceptor queue to continue not restart

## Actual Behavior
```
(defn- go-async
  "When presented with a channel as the return value of an enter function,
  wait for the channel to return a new-context (via a go block). When a new
  context is received, restart execution of the interceptor chain with that
  context.


  This function is non-blocking, returning nil immediately (a signal to halt
  further execution on this thread)."
```

As documented above, and as my logs show below, the interceptor stack restarts
```
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 20}
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 40}
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:msg "create payment", :line 119}
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:status :accepted, :line 122}
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:msg "wait for response", :line 125}
[qtp463993419-109] INFO  payment-gateway.routes.collection - {:msg "timed out, return error", :line 136}
[qtp463993419-110] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 20}
[qtp463993419-110] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 40}
```

This is a problem because a db insert happens as part of the interceptor chain,  so restarting the chain results in an error because it was inserted the first time around.

I am curious as to why the chain is restarted instead of continuing from where it went to sleep ?
Also, what is the recommended way to handle async code which involves transactions ?

## Steps to reproduce
The code waits for a response and is supposed to return a timeout error as shown below
```clojure
(catch java.util.concurrent.ExecutionException _
                    (log/info :msg "timed out, return error")
                    (assoc context :response
                           {:status 408
                            :body {:status "error"
                                   :message (format "Operation Timed out %d seconds" wait-time)}}))
```

# Environment

## Operating System (including version).
macOs Catalina, version 10.15.5

## Your current Leiningen or Boot version (`lein --version` or `boot --version`)
Leiningen 2.8.1 on Java 11.0.2 Java HotSpot(TM) 64-Bit Server VM
## Pedestal version
0.5.5

**Comment by claj** [comment]
You should be able to run an interceptor chain and then continue it even if it has asynchronous elements (of course!). The interface for going asynchronous in the interceptor chain is that you return the result from core.async `go`-macro (which returns a channel that sooner or later just returns the result of the execution of that `go`-block). 

From the `catch`-clause above it is not apparent if you are using core.async in your interceptor-chain for asynchronous operation.

If you are really using a `go`-block, there was a change in `core.async` maybe a year ago on how it handled exceptions, one has to catch them inside the same `go`-block if IIRC, because the core.async code restructuring operation cannot also maintain the correct stack for the exceptions to be thrown.

You are also using a `catch`-clause. I would consider using an error-handling interceptor, which probably makes things easier.

When you are terminating the execution because of an error with a manual `try`/`catch`, you also have to make sure to [terminate](http://pedestal.io/api/pedestal.interceptor/io.pedestal.interceptor.chain.html#var-terminate) the execution so that the interceptor chain "turns back", if this wasn't the last interceptor in the chain anyway.

Sorry if I'm over-explaining things, but it's not clear to me how your application is structured wrt to the interceptor chain.

**Comment by ddeaguiar** [comment]
@ballad89 sorry you ran into an issue. I agree with @claj that more information is required for us to understand what is going on here. One thing I can point out is that the response map you are assoc'ing to the context in your error handling does not satisfy the termination check that pedestal uses, `ring.util.response/response?`, because it does not contain a `:header` key. While this is mentioned in the [docs](http://pedestal.io/reference/response-map#_footnoteref_1), it is easy to miss. This means your interceptor chain won't immediately terminate.

I can't say if this is the cause of your problem because there's not enough information about your async usage.

**Comment by ballad89** [comment]
@claj I don't mind you over explaining.
@ddeaguiar I updated the response, and terminated same issue.
Below is the complete interceptor
```clojure
{:name ::request-payment
     :enter (fn [{:keys [request] :as context}]
              (let [{:keys [conn body-params app-config]} request
                    {:keys [referenceId]} body-params
                    kafka-config (get-in context [:request :kafka :kafka-config])
                    {:keys [kafka-root-config k-conf string-des json-des]} kafka-config
                    {{:keys [transaction-update-topic]} :topics} kafka-root-config
                    wait-id (rand-str 10)
                    wait-time 10]
                (try
                  (with-open [ws (-> (kw/new-await-result {:c-topic (:topic-name transaction-update-topic) :c-conf k-conf :k-des string-des :v-des json-des}
                                                          (fn [msg]
                                                            (let [{:keys [value]} msg]
                                                              (when (= (:referenceId value) (str referenceId))
                                                                wait-id)))
                                                          (* wait-time 1000))
                                     kw/start)]
                    (let [{:keys [ready?]} ws]
                      (when @ready?
                        (log/info :msg "create payment")
                        (dr/create-payment-request conn body-params)
                        (let [{:keys [status]} (mo/create-payment-request app-config body-params)]
                          (log/info :status  status)
                          (assoc context :response
                                 (if (= status :accepted)
                                   (let [_ (log/info :msg "wait for response")
                                         _ @(kw/process-request ws wait-id)]
                                     (log/info :msg "return")
                                     {:status 200
                                      :body {:status "success"
                                             :message "request succeeded"}})
                                   {:status 400
                                    :body {:status "error"
                                           :message "bad request"}}))))))
                  (catch java.util.concurrent.ExecutionException _
                    (log/info :msg "return error")
                    (-> context terminate
                        (assoc  :response
                                {:status 408
                                 :headers {}
                                 :body {:status "error"
                                        :message (format "Operation Timed out %d seconds" wait-time)}})))
                  (catch Throwable e
                    (log/error :ex e)
                    (assoc context :response
                           {:status 500
                            :body {:status "error"
                                   :message "Server error"}})))))}
```

process-result returns a `CompletableFuture`
```clojure
(process-request [this id]
    (let [{:keys [cache timestamps]} this
          f (CompletableFuture.)]
      (.put cache id f)
      (.put timestamps id (System/currentTimeMillis))
      f))
```

and this is how I terminate the CompletableFuture for a timeout
```clojure
(when (<= expiryTime now)
          (let [id (.getKey e)
                f (.get cache id)]
            (when f
              (.completeExceptionally f (TimeoutException. (str "Operation taking longer than " (.toMillis timeout) "ms"))))
            (remove-id cache timestamps id)))
```
As you can see I block on the main thread
The output I am seeing
```
22:12:53.153 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 21}
22:12:53.155 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 41}
22:12:53.177 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Started kafka-system", :line 689}
22:12:53.236 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "create payment", :line 119}
22:12:54.762 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:status :accepted, :line 122}
22:12:54.762 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "wait for response", :line 125}
22:13:13.184 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Stopping kafka-system", :line 696}
22:13:13.200 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Stopped kafka-system", :line 708}
22:13:13.201 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "return error", :line 135}
22:13:13.233 [qtp1655962910-287] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 21}
22:13:13.234 [qtp1655962910-287] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 41}
```
is consistent with the docs
```
(defn- go-async
  "When presented with a channel as the return value of an enter function,
  wait for the channel to return a new-context (via a go block). When a new
  context is received, restart execution of the interceptor chain with that
  context.


  This function is non-blocking, returning nil immediately (a signal to halt
  further execution on this thread)."
```
Which I contend should not be restarting the chain

**Comment by ddeaguiar** [comment]
@ballad89 where are you going async?

**Comment by ballad89** [comment]
I don't actually, the closest I come to it is blocking on a `CompletableFuture`

**Comment by claj** [comment]
Are you sure it is a restarting interceptor chain and not the client that calls the webserver that retries the request when it faces the error?

**Comment by ballad89** [comment]
@claj Yes, I am sure it restarts.
I have annotated the logs as shown below.
```
Interceptor 1
22:12:53.153 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 21}

Interceptor 2
22:12:53.155 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 41}

Interceptor 3
22:12:53.177 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Started kafka-system", :line 689}
22:12:53.236 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "create payment", :line 119}
22:12:54.762 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:status :accepted, :line 122}
22:12:54.762 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "wait for response", :line 125}
22:13:13.184 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Stopping kafka-system", :line 696}
22:13:13.200 [qtp1655962910-291] INFO  ekoh-online-kafka.consumer - {:msg "Stopped kafka-system", :line 708}
22:13:13.201 [qtp1655962910-291] INFO  payment-gateway.routes.collection - {:msg "return error", :line 135}

Interceptor 1 again
22:13:13.233 [qtp1655962910-287] INFO  payment-gateway.routes.collection - {:msg "valid public key ?", :line 21}

Interceptor 2 again
22:13:13.234 [qtp1655962910-287] INFO  payment-gateway.routes.collection - {:msg "valid referency id ?", :line 41}
```

**Comment by ddeaguiar** [comment]
@ballad89, I'm still not sure how to interpret your results and am unclear why you are correlating the observed behavior with `go-async` since you are not _going async_ by returning a channel (see [docs](http://pedestal.io/reference/interceptors#_interceptor_return_values)). Blocking on a CompletableFuture does not mean _going async_.  Having your interceptor return a core.async channel means _going async_ (see [here](https://github.com/pedestal/pedestal/blob/69f028f71e189483496715bc20ba9e330e06d8a3/interceptor/src/io/pedestal/interceptor/chain.clj#L173), for example).

I am also making a number of assumptions based on the information you provided, mainly that you are using a servlet container and that container is Jetty. Is this true?

Now, based on the interceptor you have shared, you are blocking a request processing thread. This is generally not recommended since there are a limited number of request processing threads available . Although the distinction is not made in the docs,  operations which are io bound should leverage [core.async/thread](https://clojuredocs.org/clojure.core.async/thread) as opposed to a `go` block  to offload work from the request processing threads. This will return a channel which can be used to convey the updated context once processing is complete.

By the way, there's also no need to explicitly invoke [terminate](http://pedestal.io/api/pedestal.interceptor/io.pedestal.interceptor.chain.html#var-terminate).

**Comment by ballad89** [comment]
@ddeaguiar Yes, I am using Jetty
I have wrapped everything in a thread, same result
```clojure
{:name ::request-payment
     :enter (fn [{:keys [request] :as context}]
              (let [{:keys [conn body-params app-config]} request
                    {:keys [referenceId]} body-params
                    kafka-config (get-in context [:request :kafka :kafka-config])
                    {:keys [kafka-root-config k-conf string-des json-des]} kafka-config
                    {{:keys [transaction-update-topic]} :topics} kafka-root-config
                    wait-id (rand-str 10)
                    wait-time 5]
                (async/thread
                  (try
                    (with-open [ws (-> (kw/new-await-result {:c-topic (:topic-name transaction-update-topic) :c-conf k-conf :k-des string-des :v-des json-des}
                                                            (fn [msg]
                                                              (let [{:keys [value]} msg]
                                                                (when (= (:referenceId value) (str referenceId))
                                                                  wait-id)))
                                                            (* wait-time 1000))
                                       kw/start)]
                      (let [{:keys [ready?]} ws]
                        (when @ready?
                          (log/info :msg "create payment")
                          (dr/create-payment-request conn body-params)
                          (let [{:keys [status]} (mo/create-payment-request app-config body-params)]
                            (log/info :status  status)
                            (assoc context :response
                                   (if (= status :accepted)
                                     (let [_ (log/info :msg "wait for response")
                                           _ @(kw/process-request ws wait-id)]
                                       (log/info :msg "return")
                                       {:status 200
                                        :body {:status "success"
                                               :message "request succeeded"}})
                                     {:status 400
                                      :body {:status "error"
                                             :message "bad request"}}))))))
                    (catch java.util.concurrent.ExecutionException _
                      (log/info :msg "return error")
                      (assoc context :response
                             {:status 408
                              :headers {}
                              :body {:status "error"
                                     :message (format "Operation Timed out %d seconds" wait-time)}}))
                    (catch Throwable e
                      (log/error :ex e)
                      (assoc context :response
                             {:status 500
                              :body {:status "error"
                                     :message "Server error"}}))))))}
```

**Comment by ohpauleez** [comment]
Hi all-  Jumping in here to add a bit of context.

The docstring uses the word "restart" but within the sentence, "restart with _that_ context" it is implying resume (a continuation of the interceptor chain).

If an aspect of chain execution isn't clear (or if you've found a bug), I would first check [the official docs](http://pedestal.io/reference/index) and then look at the [interceptor tests](https://github.com/pedestal/pedestal/blob/master/service/test/io/pedestal/interceptor_test.clj).  The Interceptor tests cover all known run-modes of a chain/execution and cover cases for past defects that have been addressed.  I often find it helpful to recreate my situation in this test suite because it forces me to create the most minimal reproduction of a behavior.

Another useful bit of knowledge when looking at chain execution behavior-
The interceptor chain is stored within the context.  Within every single interceptor you can log (or print) the `:io.pedestal.interceptor.chain/stack` (the interceptors that have already processed the context on :enter) and the `io.pedestal.interceptor.chain/queue` (the interceptors queued up to execute after your interceptor returns a new context).

I've been unable to recreate the behavior you're describing.  If you can link us to a minimal reproduction we'll take a look at it.  It seems like there may be a code path within your source code that's returning the original `context` instead of the updated context value.

As mentioned by @ddeaguiar - you shouldn't explicitly call `terminate` unless you're working against the interceptor chain directly (ie: you're writing a lower-level piece of Pedestal, like a chain provider).  Instead, you should allow terminator to be determined by the chain execution (and the `:io.pedestal.interceptor.chain/terminators` - collection of predicate functions that determine when termination happens).

**Comment by timewald** [comment]
Hi everyone,

There is a lot to think about in this thread...

First, as has been pointed out, this code is not "going async" because the
interceptor function is not returning a channel. As per the docstring for
go-async internal function quoted in the issue report:  "When presented
with a channel as the return value of an enter function...". That isn't
happening here. If it were, interceptor execution would pause (via a go
block) until a response context is put on the channel that was returned.
The key point here is that this is the only mechanism by which you can
pause interceptor execution until an async operation is complete. Using
other async mechanisms in your interceptor code, like an API with a
completable future, or a clojure future or a java executor service, or
anything else won't do it. You have to integrate whatever you are doing in
your interceptor with Pedestal's async channel mechanism for this to work.

The point was made that it is generally a good idea not to block the
request thread, of which there are a limited number, and that
core.async/thread should be used to do work that is IO bound (like talking
to a database, or an HTTP endpoint). However, it's a little trickier than
that. Imagine that all your requests require processing that is IO bound.
If you move all that work to other threads via core.async/thread, it is
true that you will be able to use the HTTP request threads to continue
handling other requests. However, your set of worker threads will grow
unbounded (the semantic of core.async/thread) and will ultimately slow
everything down. It may be that the limit on pending requests in Jetty will
protect you, but maybe not. So what are the alternatives here? You could
use a java executor service with a limited side thread pool to queue up
work. If you don't have too many requests at a time, and you have enough
memory, you could perhaps get the balance right.

If the thing you plan to do inside your interceptor is itself genuinely
asynchronous, then this may be less of a problem. If you can initiate true
async work and then return a channel that will later be populated when that
async work completes, then you don't need to use something like
core.async/thread. However, you need deep knowledge of the implementation
of the API you are calling to know if it is really asynchronous. If you use
the Jetty client to call a remote HTTP endpoint, you can do that truly
asynchronously. In contrast, the AWS Java SDK v1.x provides an async
interface but it uses its own thread pool to make blocking sync calls. The
2.x version has a true async interface. The point here is that people
sometimes say async meaning "not blocking the calling thread" and sometimes
meaning "not using a thread at all, or really only when necessary to
rendezvous when the work completes". You need to know which case applies
for each library you deal with.

Finally, I want to comment on databases and asynchrony. For any db work you
should be using a connection pool, and that pool should not grow unbounded.
If you spawn threads to free up request threads, you may end up with loads
of them waiting for the connection pool to have a connection available.
Overhead for per-thread context in the JVM is non trivial and so this would
likely have an impact on your performance. But if your db API is truly
asynchronous, you could kick off the work without spinning up a thread. You
would only be using the request thread long enough to initiate the work,
not for it to complete, which might be okay. However, to complete the work,
you will need to get a thread from somewhere, and if that pool is limited
in size, then you may have to wait to be able to finish a given db
interaction. If you are using transactions, then you may or may not be
holding locks of one sort or another (depending on db and what you are
doing with it), which is generally a bad idea.

All of this is to say that there is no way around it: async programming is
complex. At one extreme there are fully event driven models where
everything is async. The problem with that approach is that you almost
always have to deal with a library or some existing code that doesn't work
that way. At the other end of the spectrum is code that is just completely
synchronous and blocking. Pedestal service was designed to enable the
creation of services that can "go async" without having to do it everywhere
all the time. My best advice in this case is to "go async" only if you have
a way to limit thread growth in your code (which you get out of the box by
not going async, you are limited by the jetty request thread pool). I would
probably not make db transactions span async callbacks, but that's just my
gut instinct. I haven't built anything that way. While I know in theory it
could be made to work, I think it would turn out to be hard to reason about
and to address performance issues if they arise.

Tim-

**Comment by ddeaguiar** [comment]
@ohpauleez thanks for jumping in and providing additional context.

>  If you can link us to a minimal reproduction we'll take a look at it.
Totally agree.

@timewald, thanks for expanding on the nuances involved with the _going async_ decision. I now realize the guidance I had provided was overly terse and prescriptive. I think there's valuable wisdom in your comment which should make it's way to the docs. I'm going to make a note of that.

**Comment by hlship** [comment]
Please re-open if you can put together a reproducible test case.  You may also find that enabling debugging output for the io.pedestal.interceptor.chain namespace will be enlightening about what Pedestal is actually doing around your code.
